text/0000-export-to-oci.md

```diff
@@ -0,0 +1,410 @@
+# Meta
+[meta]: #meta
+- Name: Export to OCI format
+- Start Date: 2022-02-22
+- Author(s): Juan Bustamante (@jbustamante)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Add the capability to the `Exporter` phase to save the image to disk in [OCI Layout](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) format.
+
+# Definitions
+[definitions]: #definitions
+
+- A [Platform](https://buildpacks.io/docs/concepts/components/platform/) uses a lifecycle, Buildpacks (packaged in a builder), and application source code to produce an OCI image.
+- A [Lifecycle](https://buildpacks.io/docs/concepts/components/lifecycle/) orchestrates Buildpacks execution, then assembles the resulting artifacts into a final app image.
+- A **Daemon** is a service, popularized by Docker, for downloading container images, and executing and managing containers from those images.
+- A **Registry** is a long-running service used for storing and retrieving container images.
+- An **image reference** refers to either a tag reference or digest reference.
+- A **tag reference** refers to an identifier of form `<registry>/<repo>:<tag>` which locates an image manifest in an [OCI Distribution Specification](https://github.com/opencontainers/distribution-spec/blob/master/spec.md) compliant registry.
+- A **digest reference** refers to a [content addressable](https://en.wikipedia.org/wiki/Content-addressable_storage) identifier of form `<registry>/<repo>@<digest>` which locates an image manifest in an [OCI Distribution Specification](https://github.com/opencontainers/distribution-spec/blob/master/spec.md) compliant registry.
+- A **image Manifest** provides a configuration and set of layers for a single container image for a specific architecture and operating system.
+- The **layer diffID** is the hash of the uncompressed layer
+- The **layer digest** is the hash of the compressed layer.
+- An [OCI Image Layout](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) is the directory structure for OCI content-addressable blobs and [location-addressable](https://en.wikipedia.org/wiki/Content-addressable_storage#Content-addressed_vs._location-addressed) references.
+
+# Motivation
+[motivation]: #motivation
+
+### Why should we do this?
```

Comment:
(This is maybe not the correct place for this comment, feel free to delete it if it's outside the scope of this PR)

This RFC and its associated implementation is really huge for end users, even if it may not pick up that much traction yet.

Right now, the current implementation offer 2 mains way to handle the buildpack resulting image:
* Set up your daemon to be able to push to  a registry
* Use another tool (any kind of docker client) to transform it into an artifact ([usually tar ball](https://github.com/buildpacks/lifecycle/issues/423))

Which is fine and dandy on your own machine but become quite a nightmare once implemented on Corporate Entreprise™ CI/CD. For the rest of my post, assume I'm moslty talking about containerized CI/CD which are becoming ubiquitous.

**Solution 1: Set up your daemon to be able to push to  a registry**
During the build phase of projects, organisation want to be able to segregate these environnements from the rest. They tend to be untrustable by nature (way too many thing are required to build projects nowadays). 
As most companies tend to be siloed, analysis tools are not available at build stage, so no analysis of the resulting image is possible before being pushed. 
For the same reason, build fragment tend to be maintained by developers with less security knowledge than the ones maintained by infrastructure/security teams.
You need to set up your "build" daemon to push to production registries [with patch that may end up exposing secrets really fast](https://stackoverflow.com/questions/46418652/docker-pull-returns-denied-access-forbidden-from-private-gitlab-registry/46422186#46422186). 
Overral allowing "build daemons" to push to registries end up being problematic really fast.

**Solution 2: Use another tool (any kind of docker client) to transform it into an artifact**
That's another can of worm. Most base build image don't come with a docker client by default. Which is great because it reduce the surface of attack drastically. But when you suddenly need one, it becomes quite a challenge to do that in a secure way:
* Creating base image with build tools  AND docker client. That's an organisational challenge for most company. These images tend to become unsecured quite fast. Either by being not up to date or because they become God Like Image with access to everything everywhere. We are looking at buildpack as a way to reduce our surface of attack drastically both at runtime and build time.
* Using 3rd party client/plugin to do that for you. [Exemple for maven plugin](https://github.com/fabric8io/docker-maven-plugin). That's a lot of added complexity and risk. 

Being able to generate a local artifact to be pushed outside the build time to "analyse steps" and "registries handling steps" maintained by teams better equiped to handle security and infrastructure would be really great.

So I wholeheartedly approve this PR. Thumb up, emoji heart, keep up the good work 😄.

Reply:
@Choobz Thanks a lot for your feedback, I really appreciate it! and the use cases you mentioned are very valuable! I will try to incorporate them in the RFC